TombEngine/Objects/TR5/Entity/tr5_twogunlaser.cpp

```diff
@@ -0,0 +1,475 @@
+#include "people.h"
+#include "Specific/setup.h"
+#include "Game/effects/spark.h"
+#include "Game/effects/lightning.h"
+
+using namespace TEN::Effects::Items;
+using namespace TEN::Effects::Lightning;
+using namespace TEN::Math::Random;
+using namespace TEN::Effects::Spark;
+
+namespace TEN::Entities::Creatures::TR5
+{
+	int TWGunMesh[2] = { 8, 5 };
+
+	const Vector3i TWGunPositions[2] =
+	{
+		Vector3i(0, 230, 40),
+		Vector3i(8, 230, 40),
+	};
+
+	const auto LaserguardHead = BiteInfo(Vector3(0.0f, -200.0f, 0.0f), 2);
+
+	enum TwogunState
+	{
+		TWOGUN_STATE_EMPTY = -1,
+		TWOGUN_STATE_IDLE = 1,
+		TWOGUN_STATE_WALK = 2,
+		TWOGUN_STATE_SHOOT_RIGHT = 3,
+		TWOGUN_STATE_SHOOT_LEFT = 4,
+		TWOGUN_STATE_IDLE_TO_AIM = 5,
+		TWOGUN_STATE_SHOOT_BOTH = 6,
+		TWOGUN_STATE_DEATH = 7,
+		TWOGUN_STATE_TURN_180 = 8,
+		TWOGUN_STATE_MISSFIRE = 9,
+		TWOGUN_STATE_GUN_BLOCKAGE = 10,
+		TWOGUN_STATE_FALLSTART = 11,
+		TWOGUN_STATE_FALLLOOP = 12,
+		TWOGUN_STATE_FALLDEATH = 13
+	};
+
+	enum TwogunAnim
+	{
+		TWOGUN_ANIM_WALK = 0,
+		TWOGUN_ANIM_SHOOT_RIGHT = 1,
+		TWOGUN_ANIM_SHOOT_LEFT = 2,
+		TWOGUN_ANIM_DEATH = 3,
+		TWOGUN_ANIM_AIM_TO_WALK = 4,
+		TWOGUN_ANIM_WALK_AIM_TO_IDLE = 5,
+		TWOGUN_ANIM_IDLE = 6,
+		TWOGUN_ANIM_IDLE_TO_AIM = 7,
+		TWOGUN_ANIM_AIM = 8,
+		TWOGUN_ANIM_SHOOT_BOTH = 9,
+		TWOGUN_ANIM_AIM_TO_IDLE = 10,
+		TWOGUN_ANIM_TURN_180 = 11,
+		TWOGUN_ANIM_MISSFIRE = 12,
+		TWOGUN_ANIM_GUN_BLOCKAGE = 13,
+		TWOGUN_ANIM_FALLSTART = 14,
+		TWOGUN_ANIM_FALLLOOP = 15,
+		TWOGUN_ANIM_FALLDEATH = 16
+	};
+
+	void InitialiseTwogun(short itemNumber)
+	{
+		auto* item = &g_Level.Items[itemNumber];
+
+		ClearItem(itemNumber);
+		SetAnimation(item, 6);
+	}
+
+	void FireTwogunWeapon(short itemNumber, short LeftRight, short plasma)
+	{
+		auto* item = &g_Level.Items[itemNumber];
+
+		auto pos1 = GetJointPosition(item, TWGunMesh[LeftRight], TWGunPositions[LeftRight]);
+		auto pos2 = GetJointPosition(item, TWGunMesh[LeftRight], Vector3i(TWGunPositions[LeftRight].x, TWGunPositions[LeftRight].y + 4096, TWGunPositions[LeftRight].z));
+		auto orient = Geometry::GetOrientToPoint(pos1.ToVector3(), pos2.ToVector3());
+
+		if (!plasma)
+		{
+			TriggerLaserBeam(pos1, pos2, LeftRight);
+
+			item->ItemFlags[LeftRight] = 16;
+			TriggerTwogunPlasma(pos1, Pose(pos1.ToVector3(), orient), 16);
+			TriggerTwogunPlasma(pos1, Pose(pos1.ToVector3(), orient), 16);
+			TriggerTwogunPlasma(pos1, Pose(pos1.ToVector3(), orient), 16);
+
+			Vector3i hitPos;
+			MESH_INFO* hitMesh = nullptr;
+
+			GameVector start = GameVector(pos1, 0);
+			GameVector end = GameVector(pos2, 0);
+			start.RoomNumber = item->RoomNumber;
+
+			if (ObjectOnLOS2(&start, &end, &hitPos, &hitMesh, ID_LARA) == GetLaraInfo(LaraItem)->ItemNumber)
+			{
+				if (LaraItem->HitPoints < 501)
+				{
+					ItemCustomBurn(LaraItem, Vector3(0.2f, 0.4f, 1.0f), Vector3(0.2f, 0.3f, 0.8f), 1 * FPS);
+					DoDamage(LaraItem, INT_MAX);
+				}
+				else
+					DoDamage(LaraItem, 250);
+			}
+		}
+	
+			return;
```

Comment:
What is this? Why `return` is here? Next code line (`TriggerTwogunPlasma`) is unreachable!